### PR TITLE
DB-11903 Materialize updates into all values

### PR DESCRIPTION
**Pull request by dgomezferro**
## Short Description
Merge existing data with update to materialize all values into the beginning of the MVCC history for each row.

## Long Description
Currently we only write what has changed in each update, but in some cases this causes performance degradation, for instance when updating repeatedly a single column we end up generating a large MVCC history that we have to scan over when we want to reconstruct the full row.

This change materializes the full row for each update, so that we always have the latest values at the beginning of the MVCC history.

Caveats:
- During WW conflict if the latest value is rolledback we don't perform the materialization
- This adds some overhead (both CPU / Memory and storage)

## How to test
Repeated updates to a single column shouldn't deteriorate full row scans performance anymore.

### CODE REVIEW

**Comment by dgomezferro** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by msirek — APPROVED** [review]

**Review by hatyo — APPROVED** [review]
Nice work, I really like the tests you added, thanks for that!

**Review by arnaud-splice — CHANGES_REQUESTED** [review]
+1 once Youssef comments were addressed

**Review by carolp-503 — APPROVED** [review]
+1

File: splice_access_api/src/main/java/com/splicemachine/storage/util/UpdateUtils.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    public static KVPair updateFromWrite(KVPair mutation) {
+        return extractWriteFromUpdate(mutation, KVPair.Type.UPDATE);
+    }
+
+    /**
+     * Returns the first half of a bit set
+     */
+    public static BitSet halveSet(BitSet bitSet) {
+        BitSet result = new BitSet(bitSet.capacity()/2);
```
[the remark this answers is not in the document]
Comment (dgomezferro) [reply splice_access_api/src/main/java/com/splicemachine/storage/util/UpdateUtils.java:94]: If I understand it correctly it expects the maximum index/value to be set on the BitSet, not the maximum number of set bits.

For instance:
```
BitSet s = new BitSet(64);
s.capacity(); // 64
s.set(100); // 100 > 64, needs to resize
s.capacity(); // 320
```

File: splice_access_api/src/main/java/com/splicemachine/storage/util/UpdateUtils.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    public static KVPair updateFromWrite(KVPair mutation) {
+        return extractWriteFromUpdate(mutation, KVPair.Type.UPDATE);
+    }
+
+    /**
+     * Returns the first half of a bit set
+     */
+    public static BitSet halveSet(BitSet bitSet) {
```
[the remark this answers is not in the document]
Comment (dgomezferro) [reply splice_access_api/src/main/java/com/splicemachine/storage/util/UpdateUtils.java:93]: Yes, I should have documented this, but this method expects the BitSet to be the concatenation of two copies of an original BitSet, so in your case input should be something like `[0,0,1,0,1,0,0,1,0,1]`:
```
BitSet bitSet = new BitSet(10); // bitSet=[0,0,0,0,0,0,0,0,0,0]
bitSet.set(3);
bitSet.set(5);
bitSet.set(8);
bitSet.set(10);                 // bitSet=[0,0,1,0,1,0,0,1,0,1]
BitSet half = halveSet(bitSet); // half=  [0,0,1,0,1]
System.out.println(half); // result is: {3,5}
```
I'll improve the documentation here, thanks!

File: splice_access_api/src/main/java/com/splicemachine/storage/util/UpdateUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (hatyo) [reply splice_access_api/src/main/java/com/splicemachine/storage/util/UpdateUtils.java:97]: You're right, I was confused about it, thanks for checking it and correcting my understanding.

**Review by arnaud-splice — APPROVED** [review]

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**